malloc_free_check/malloc_free_check_extension.py

```diff
@@ -0,0 +1,135 @@
+"""
+Undo Automation extension module for tracking calls to malloc() and free() and checking for
+leaked memory.
+
+This script only support the x86-64 architecture.
+
+Contributors: Chris Croft-White, Magne Hov
+"""
+
+import collections
+import re
+
+import gdb
+
+from undodb.debugger_extensions import udb
+from undodb.debugger_extensions.debugger_io import redirect_to_launcher_output
+
+
+def leak_check() -> int:
+    """
+    Implements breakpoints and stops on all calls to malloc() and free(), capturing the
+    timestamp, size and returned pointer for malloc(), then confirms the address pointer is later
+    seen in a free() call.
+
+    If a subsequent free() is not seen, then at the end of execution, output the timestamp and
+    details of the memory which was never freed.
+
+    Returns the number of unmatched allocations found.
+    """
+    # Set a breakpoint for the specified function.
+    gdb.Breakpoint("malloc")
+    gdb.Breakpoint("free")
+
+    # Declare allocations dictionary structure.
+    allocations = collections.OrderedDict()
+
+    # Do "continue" until we have gone through the whole recording, potentially
+    # hitting the breakpoints several times.
+    end_of_time = udb.get_event_log_extent().end
+    while True:
+        gdb.execute("continue")
+
+        # Rather than having the check directly in the while condition we have
+        # it here as we don't want to print the backtrace when we hit the end of
+        # the recording but only when we stop at a breakpoint.
+        if udb.time.get().bbcount >= end_of_time:
+            break
+
+        # Use the $PC output to get the symbol and idenfity whether execution has stopped
+        # at a malloc() or free() call.
+        mypc = format(gdb.parse_and_eval("$pc"))
+        if re.search("malloc", mypc):
+            # In malloc(), set a FinishBreakpoint to capture the pointer returned later.
+            mfbp = gdb.FinishBreakpoint()
+
+            # For now, capture the timestamp and size of memory requested.
+            time = udb.time.get()
+            size = int(gdb.parse_and_eval("$rdi"))
+
+            gdb.execute("continue")
+
+            # Should stop at the finish breakpoint, so capture the pointer.
+            assert mfbp.return_value is not None, "Expected to see a return value."
+            addr = int(mfbp.return_value)
+
+            if addr:
+                # Store details in the dictionary.
+                allocations[hex(addr)] = time, size
```

Comment:
It would be cleanear if the address was stored as an integer rather than its string representation.